py/sentry_streams/flink/flink_fn_translator.py

```diff
@@ -0,0 +1,114 @@
+from datetime import timedelta
+from typing import Any
+
+from pyflink.common import Time, Types
+from pyflink.datastream.functions import AggregateFunction, KeySelector
+from pyflink.datastream.window import (
+    CountSlidingWindowAssigner,
+    CountTrigger,
+    CountTumblingWindowAssigner,
+)
+from pyflink.datastream.window import EventTimeTrigger as FlinkEventTimeTrigger
+from pyflink.datastream.window import (
+    SlidingEventTimeWindows,
+)
+from pyflink.datastream.window import Trigger as FlinkTrigger
+from pyflink.datastream.window import (
+    TumblingEventTimeWindows,
+    WindowAssigner,
+)
+
+from sentry_streams.user_functions.function_template import Accumulator, GroupBy
+from sentry_streams.window import (
+    CountingTrigger,
+    EventTimeTrigger,
+    SlidingCountWindow,
+    SlidingEventTimeWindow,
+    Trigger,
+    TumblingCountWindow,
+    TumblingEventTimeWindow,
+    Window,
+)
+
+FLINK_TYPE_MAP = {tuple[str, int]: Types.TUPLE([Types.STRING(), Types.INT()]), str: Types.STRING()}
+
+
+class FlinkAggregate(AggregateFunction):
+
+    def __init__(self, acc: Accumulator) -> None:
+        self.acc = acc
+
+    def create_accumulator(self) -> Any:
+        return self.acc.create()
+
+    def add(self, value: Any, accumulator: Any) -> Any:
+        return self.acc.add(accumulator, value)
+
+    def get_result(self, accumulator: Any) -> Any:
+        return self.acc.get_output(accumulator)
+
+    def merge(self, acc_a: Any, acc_b: Any) -> Any:
+        return self.acc.merge(acc_a, acc_b)
+
+
+class FlinkGroupBy(KeySelector):
+
+    def __init__(self, group_by: GroupBy) -> None:
+        self.group_by = group_by
+
+    def get_key(self, value: Any) -> Any:
+        return self.group_by.get_group_by_key(value)
+
+
+def to_flink_time(timestamp: timedelta) -> Time:
+    """
+    Convert a timedelta object to a Time type representation in milliseconds.
+    """
+
+    total_milliseconds: int = int(timestamp.total_seconds() * 1000)
+
+    flink_time: Time = Time.milliseconds(total_milliseconds)
+    return flink_time
+
+
+class FlinkWindows:
```

Comment:
I think this translation needs some work. 

Right now, the API exposes only a few specific window, trigger combinations that a user can pick from. One benefit of this is that we can just use some of the `WindowAssigners` that Flink offers out of the box. Down the road though, this translation will need to be more flexible. 

Flink exposes a `WindowAssigner` interface which we could try to implement as a translation mechanism. This is a possibility, will just require a little more understanding of Flink's internals. 

Open to any other ideas as well.

Reply:
Exposing a subset of the primitives is ok. Actually it is a desirable behavior for a platform:
- We can be opinionated on the data model and the capabilities we want to expose rather than simply inheriting everything the runtime underneath supports. This allows us to focus on what is important for the application and allows for maintainability.
- We have an easier time in supporting multiple runtimes.
- The value of the platform is often to put the right restrictions to guarantee we follow an architecture, which allows for consistency in how we solve problems.

Obviously finding the right constraints and the right primitives to expose is the hard part. At this point what you provided is more than enough.